packages/context/src/binding.ts

```diff
@@ -503,9 +504,18 @@ export class Binding<T = BoundValue> {
       debug('Bind %s to class %s', this.key, ctor.name);
     }
     this._type = BindingType.CLASS;
-    this._setValueGetter((ctx, options) =>
-      instantiateClass(ctor, ctx, options.session),
-    );
+    this._setValueGetter((ctx, options) => {
+      const instOrPromise = instantiateClass(ctor, ctx, options.session);
+      if (!options.asProxyWithInterceptors) return instOrPromise;
+      return transformValueOrPromise(instOrPromise, inst => {
+        if (typeof inst !== 'object') return inst;
+        return (createProxyWithInterceptors(
+          // Cast inst from `T` to `object`
+          (inst as unknown) as object,
```

Comment:
Why is TypeScript not able to infer that `inst` must be an `object`, considering the `typeof inst !== 'object'` typecheck above? Can you try to rewrite this code in such way that no explicit casts are needed? 

Would this work?

```ts
if (typeof inst !== 'object') return inst;
const obj: object = inst;
const proxy = createProxyWithInterceptors(obj, ctx);
return (proxy as unknown) as T;
```

Also I see that `creatzeProxyWithInterceptors` is already parameterized, why cannot we leverage that generic parameter here? If we cannot leverage it, then what's the point of having that parameter in the first place?

```ts
return createProxyWithInterCeptors<T>(obj, ctx);
```

Is the problem perhaps in the fact that our current type descriptions allow `T` to be any value, not only objects? IIUC, it's possible to create `new Binding<number>().toClass(/*...*/)`, which does not look right to me. Perhaps we need to restrict the types for which `toClass` is available, something along the lines of `T extends object`. Can we change `Constructor<T>` alias for that? I am not sure if that would be considered as a breaking change or not.

Reply:
TypeScript cannot infer `inst` as `object` after `if (typeof inst !== 'object') return inst;`.